File: modules/map3d/android/src/main/java/expo/modules/map3d/Map3DModule.kt
package expo.modules.map3d

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class Map3DModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Map3D")

    View(ExpoMap3DView::class) {
      Events("onStatusChange")

      Prop("center") { view, center: Map<String, Double> ->
        val lat = center["latitude"] ?: return@Prop
        val lng = center["longitude"] ?: return@Prop
        view.setCenter(lat, lng)
      }

      Prop("mapMode") { view, mode: String? ->
        view.setMapMode(mode ?: "HYBRID")
      }

      Prop("markerPosition") { view, position: Map<String, Double>? ->
        val lat = position?.get("latitude") ?: return@Prop
        val lng = position["longitude"] ?: return@Prop
        view.setMarkerPosition(lat, lng)
      }

      Prop("routeCoordinates") { view, points: List<Map<String, Double>>? ->
        view.setRouteCoordinates(points ?: emptyList())
      }

      Function("rotateCamera") { view: ExpoMap3DView, deltaDeg: Double ->
        view.rotateCamera(deltaDeg)
      }

      Function("tiltCamera") { view: ExpoMap3DView, deltaDeg: Double ->
        view.tiltCamera(deltaDeg)
      }
    }
  }
}

File: modules/map3d/ios/Map3DModule.swift
import ExpoModulesCore

public class Map3DModule: Module {
  public func definition() -> ModuleDefinition {
    Name("Map3D")

    View(ExpoMap3DView.self) {
      Events("onStatusChange")

      Prop("center") { (view: ExpoMap3DView, center: [String: Double]) in
        guard let lat = center["latitude"], let lng = center["longitude"] else { return }
        view.setCenter(latitude: lat, longitude: lng)
      }

      Prop("mapMode") { (view: ExpoMap3DView, mode: String?) in
        view.setMapMode(mode ?? "HYBRID")
      }

      Prop("markerPosition") { (view: ExpoMap3DView, position: [String: Double]?) in
        guard let lat = position?["latitude"], let lng = position?["longitude"] else { return }
        view.setMarkerPosition(latitude: lat, longitude: lng)
      }

      Prop("routeCoordinates") { (view: ExpoMap3DView, points: [[String: Double]]?) in
        view.setRouteCoordinates(points ?? [])
      }

      Function("rotateCamera") { (view: ExpoMap3DView, deltaDeg: Double) in
        view.rotateCamera(deltaDeg: deltaDeg)
      }

      Function("tiltCamera") { (view: ExpoMap3DView, deltaDeg: Double) in
        view.tiltCamera(deltaDeg: deltaDeg)
      }
    }
  }
}
